docs/tutorial/write-your-first-kubernetes-charm-for-a-flask-app.rst

```diff
@@ -0,0 +1,645 @@
+.. _write-your-first-kubernetes-charm-for-a-flask-app:
+
+Write your first Kubernetes charm for a Flask app
+=================================================
+
+Imagine you have a Flask application backed up by a database
+such as PostgreSQL and need to deploy it. In a traditional setup,
+this can be quite a challenge, but with Charmcraft you'll find
+yourself packaging and deploying your Flask application in no time.
+Let's get started!
+
+In this tutorial we will build a Kubernetes charm for a Flask
+application using Charmcraft, so we can have a Flask application
+up and running with Juju.
+
+This tutorial should take 90 minutes for you to complete.
+
+.. note::
+    If you're new to the charming world: Flask applications are
+    specifically supported with a coordinated pair of profiles
+    for an OCI container image (**rock**) and corresponding
+    packaged software (**charm**) that allow for the application
+    to be deployed, integrated and operated on a Kubernetes
+    cluster with the Juju orchestration engine.
+
+What you'll need
+----------------
+
+- A workstation, e.g., a laptop, with amd64 or arm64 architecture which
+  has sufficient resources to launch a virtual machine with 4 CPUs,
+  4 GB RAM, and a 50 GB disk.
+- Familiarity with Linux.
+
+What you'll do
+--------------
+
+Create a Flask application. Use that to create a rock with
+``rockcraft``. Use that to create a charm with ``charmcraft``. Use that
+to test, deploy, configure, etc., your Flask application on a local
+Kubernetes cloud, ``microk8s``, with ``juju``. All of that multiple
+times, mimicking a real development process.
+
+.. important::
+
+    Should you get stuck or notice issues, please get in touch on
+    `Matrix <https://matrix.to/#/#12-factor-charms:ubuntu.com>`_ or
+    `Discourse <https://discourse.charmhub.io/>`_
+
+Set things up
+-------------
+
+.. include:: /reuse/tutorial/setup_stable.rst
+.. |12FactorApp| replace:: Flask
+
+Let's create a new directory for this tutorial and change into it:
+
+.. code-block:: bash
+
+    mkdir flask-hello-world
+    cd flask-hello-world
+
+Finally, install ``python3-venv`` and create a virtual environment:
+
+.. literalinclude:: code/flask/task.yaml
+    :language: bash
+    :start-after: [docs:create-venv]
+    :end-before: [docs:create-venv-end]
+    :dedent: 2
+
+Create the Flask application
+----------------------------
+
+Let's start by creating the "Hello, world" Flask application that
+will be used for this tutorial.
+
+Create a ``requirements.txt`` file using ``touch requirements.txt``.
+Then, open the file in a text editor using ``nano requirements.txt``,
+copy the following text into it and then save the file:
+
+.. literalinclude:: code/flask/requirements.txt
+
+.. note::
+
+   The ``psycopg2-binary`` package is needed so the Flask application can
```

Comment:
I understand why the tutorial is this way, but at some point it'd probably be worth documenting uhh... somewhere (no idea where is the right place except that it's definitely out of scope here) how to do the "preferred" method of getting psycopg2 from source and including relevant Ubuntu packages in the rock.